lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -1162,28 +1162,41 @@ class PullbackCloner::Implementation final
       // Accumulate adjoint for the `struct_extract` operand.
       auto av = getAdjointValue(bb, sei);
       switch (av.getKind()) {
-      case AdjointValueKind::Zero:
-        addAdjointValue(bb, sei->getOperand(),
-                        makeZeroAdjointValue(tangentVectorSILTy), loc);
+      case AdjointValueKind::Concrete: {
+        // Materialize adj[x]
+        auto *adjBaseAlloc = builder.createAllocStack(loc, tangentVectorSILTy);
+        materializeAdjointIndirect(makeZeroAdjointValue(tangentVectorSILTy),
+                                   adjBaseAlloc, loc);
+
+        // Materialize adj[y]
+        auto adjElt = materializeAdjointDirect(av, loc);
+        // Copy `adjElt` so we can get a value with owned ownership
+        // semantics, required for using `adjElt` in the store instruction.
+        auto adjEltCopy = builder.emitCopyValueOperation(loc, adjElt);
+        auto *adjEltAlloc = builder.createAllocStack(loc, adjElt->getType());
+        builder.emitStoreValueOperation(loc, adjEltCopy, adjEltAlloc,
+                                        StoreOwnershipQualifier::Init);
+
+        // Get adj[x].#field`
+        auto *adjBaseEltAddr =
+            builder.createStructElementAddr(loc, adjBaseAlloc, tanField);
+
+        // adj[x].#field` += adj[y]
+        builder.emitInPlaceAdd(loc, adjBaseEltAddr, adjEltAlloc);
+
+        auto adjBaseConcrete = recordTemporary(builder.emitLoadValueOperation(
+            loc, adjBaseAlloc, LoadOwnershipQualifier::Take));
+        builder.createDestroyAddr(loc, adjEltAlloc);
+        builder.createDeallocStack(loc, adjEltAlloc);
+        builder.createDeallocStack(loc, adjBaseAlloc);
+        setAdjointValue(bb, sei->getOperand(),
+                        makeConcreteAdjointValue(adjBaseConcrete));
         break;
-      case AdjointValueKind::Concrete:
+      }
+      case AdjointValueKind::Zero:
       case AdjointValueKind::Aggregate: {
-        SmallVector<AdjointValue, 8> eltVals;
-        for (auto *field : tangentVectorDecl->getStoredProperties()) {
-          if (field == tanField) {
-            eltVals.push_back(av);
-          } else {
-            auto substMap = tangentVectorTy->getMemberSubstitutionMap(
-                field->getModuleContext(), field);
-            auto fieldTy = field->getType().subst(substMap);
-            auto fieldSILTy = getTypeLowering(fieldTy).getLoweredType();
-            assert(fieldSILTy.isObject());
-            eltVals.push_back(makeZeroAdjointValue(fieldSILTy));
-          }
-        }
-        addAdjointValue(bb, sei->getOperand(),
-                        makeAggregateAdjointValue(tangentVectorSILTy, eltVals),
-                        loc);
+        llvm_unreachable("Adjoint value kind of the input to `struct_extract` "
```

Comment:
With this, you are now forcing all `struct_extract` to materialize the adjoint. To me it seems unnecessary and may lead to performance issues for common cases where there is no non-differentiable fields of a `TangentVector`.

Can we use a type-driven approach instead?  It is still fine to create an `Aggregate` adjoint value or use the `struct` instruction, so long as the following conditions are met:
1. `TangentVector` is defined in the same compilation unit of differentiation.
2. All of `TangentVector`’s stored properties have a same-name differentiable property in the original type.

Or better, you could create a new `AdjointValueKind` called `AddElement`.  When the above conditions aren't met, just create a `AddElement(adj[x], #field, adj[y])` adjoint value. This way the whole thing can be kept symbolic and won't be materialized unless it's absolutely necessary.

Reply:
Thanks for reviewing this Richard! I have a follow up question.

> With this, you are now forcing all struct_extract to materialize the adjoint. To me it seems unnecessary and may lead to performance issues for common cases where there is no non-differentiable fields of a TangentVector.

Aren't all adjoints eventually going to materialize, before the end of the differentiation pass, given the differentiation pass is run once and atomically(right?) with no other optimization passes run in between that may perform some peephole optimizations?

Reply:
> Aren't all adjoints eventually going to materialize

No, the purpose of the `AdjointValue` data structure is to prevent eager materialization and try to fold the data structure as much as possible. This is why we have special handling of different `AdjointValueKind`s for most instructions. There's a lot of cases of creating a struct of all zeros but one element, then using just one element from it in subsequent code.

Reply:
That makes sense Richard. I also discussed this with Brad, Anton and others and it seems like the right direction, at the moment. Let me make the changes you suggested and get a new revision out.

Reply:
@rxwei I was wondering something about this new adjoint kind we need to create. Could it just be a closure that simply returns a `SILValue`? I'm wondering if a new adjoint kind like `AddElement(adj[x], #field, adj[y])` is too specific and perhaps we can create something more general that works for all adjoints that may need some custom materialization logic?

Reply:
Another follow-up question. 

Would you be able to comment on what the representation, in the actual implementation of the new adjoint kind you're suggesting (`AddElement(adj[x], #field, adj[y])`), should look like? The 2 I can think of are:
1. Store the original `struct_extract` or `tuple_extract` instruction.
2. Store the adjoints of `x` and `y` and the `VarDecl` for the field in `y = struct_extract x, #field`

Reply:
You need to be able to symbolically evaluate the adjoint in case it would be needed for materialization. Note that both `x` and `y` here could also be symbolic adjoints. So, the `2.` looks like a way to go for me.